src/main/scala/scala/collection/decorators/IteratorDecorator.scala

```diff
@@ -72,4 +72,63 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
     loop(immutable.List.empty)
   }
 
+  /**
+    * Constructs an iterator where consecutive elements are accumulated as
+    * long as the output of f for each element doesn't change.
+    * <pre>
+    * Vector(1,2,2,3,3,3,2,2)
+    * .iterator
+    * .splitBy(identity)
+    * .toList
+    * </pre>
+    * produces
+    * <pre>
+    * List(Seq(1),
+    * Seq(2,2),
+    * Seq(3,3,3),
+    * Seq(2,2))
+    * </pre>
+    *
+    * @param f the function to compute a key for an element
+    * @tparam K the type of the computed key
+    * @return an iterator of sequences of the consecutive elements with the
+    *         same key in the original iterator
+    */
+  def splitBy[K](f: A => K): Iterator[immutable.Seq[A]] =
+    new AbstractIterator[immutable.Seq[A]] {
+      private var hd: A = _
+      private var hdDefined: Boolean = false
+
+      override def hasNext: Boolean = hdDefined || `this`.hasNext
+
+      override def next(): immutable.Seq[A] = {
+        if (hasNext) {
+          val seq = Vector.newBuilder[A]
+          if (hdDefined) {
+            seq += hd
+          } else {
+            hd = `this`.next()
+            hdDefined = true
+            seq += hd
+          }
+          var hadSameKey = true
+          while (`this`.hasNext && hadSameKey) {
+            val el = `this`.next()
+            hdDefined = true
+            if (f(el) == f(hd)) {
```

Comment:
Same comment here about `f(hd)`

Reply:
That wouldn't work, `hd` needs to be mutable to honor the iterator contract. It is mutated on line 123. 

The best I could come up with is adding another mutable variable `hdKey` which is only updated when the key actually changes. 
I have added a commit to demonstrate it. I feel it does damage the readability but I understand it can potentially have a huge performance impact.